pkg/alibabacloud/api/api.go

```diff
@@ -0,0 +1,501 @@
+type MetricsAPI interface {
+	helpers.MetricsAPI
+	ObserveAPICall(call, status string, duration float64)
+}
+
+// NewClient create the client
+func NewClient(vpcClient *vpc.Client, client *ecs.Client, metrics MetricsAPI, rateLimit float64, burst int, filters map[string]string) *Client {
+	return &Client{
+		vpcClient:  vpcClient,
+		ecsClient:  client,
+		limiter:    helpers.NewApiLimiter(metrics, rateLimit, burst),
+		metricsAPI: metrics,
+		filters:    filters,
+	}
+}
+
+// GetInstances returns the list of all instances including their ENIs as instanceMap
+func (c *Client) GetInstances(ctx context.Context, vpcs ipamTypes.VirtualNetworkMap, subnets ipamTypes.SubnetMap) (*ipamTypes.InstanceMap, error) {
+	instances := ipamTypes.NewInstanceMap()
+
+	networkInterfaceSets, err := c.describeNetworkInterfaces(ctx, subnets)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, iface := range networkInterfaceSets {
+		id, eni, err := parseENI(&iface, vpcs, subnets)
+		if err != nil {
+			return nil, err
+		}
+
+		instances.Update(id, ipamTypes.InterfaceRevision{
+			Resource: eni,
+		})
+	}
+	return instances, nil
+}
+
+// GetVSwitches returns all ecs vSwitches as a subnetMap
+func (c *Client) GetVSwitches(ctx context.Context) (ipamTypes.SubnetMap, error) {
+	var result ipamTypes.SubnetMap
+	vsws := []string{}
+	for i := 1; ; {
+		req := vpc.CreateDescribeVSwitchesRequest()
+		req.PageNumber = requests.NewInteger(i)
+		req.PageSize = requests.NewInteger(50)
+		req.VpcId = c.filters[VPCID]
+		c.limiter.Limit(ctx, "DescribeVSwitches")
+		resp, err := c.vpcClient.DescribeVSwitches(req)
+		if err != nil {
+			return nil, err
+		}
+		if len(resp.VSwitches.VSwitch) == 0 {
+			break
+		}
+		if result == nil {
+			result = make(ipamTypes.SubnetMap, resp.TotalCount)
+		}
+
+		for _, v := range resp.VSwitches.VSwitch {
+			_, ipnet, err := net.ParseCIDR(v.CidrBlock)
+			if err != nil {
+				return nil, err
+			}
+			result[v.VSwitchId] = &ipamTypes.Subnet{
+				ID:                 v.VSwitchId,
+				Name:               v.VSwitchName,
+				CIDR:               cidr.NewCIDR(ipnet),
+				AvailabilityZone:   v.ZoneId,
+				VirtualNetworkID:   v.VpcId,
+				AvailableAddresses: int(v.AvailableIpAddressCount),
+				Tags:               map[string]string{},
+			}
+			vsws = append(vsws, v.VSwitchId)
+		}
+		if resp.TotalCount < resp.PageNumber*resp.PageSize {
+			break
+		}
+		i++
+	}
+
+	for i := 0; i <= (len(vsws)-1)/MaxListByTagSize; i++ {
+		var ids []string
+
+		tail := min((i+1)*MaxListByTagSize, len(vsws))
+		ids = vsws[i*MaxListByTagSize : tail]
+
+		req := vpc.CreateListTagResourcesRequest()
+		req.ResourceType = "VSWITCH"
+		req.ResourceId = &ids
+		c.limiter.Limit(ctx, "ListTagResources")
+		resp, err := c.vpcClient.ListTagResources(req)
+		if err != nil {
+			return nil, err
+		}
+		for _, tagRes := range resp.TagResources.TagResource {
+			subnet, ok := result[tagRes.ResourceId]
+			if !ok {
+				log.Errorf("can not found vSwitch id %s", tagRes.ResourceId)
```

Comment:
Does error level severity make sense? I'm not sure I fully understand what would happen if this case is hit. Could you describe a bit what would happen if we cannot find the vSwitch ID? How likely is this error? 

One way to think about logging msgs to the user is what are they supposed to do when they see this msg? If they cannot, then the next question is how can we provide enough context in this log msg so that it's meaningful to us when debugging.

```suggestion
				log.WithField(logfields.vSwitchID, tagRes.ResourceID).Warn("vSwitch ID was not found")
```

Reply:
it's expected ,should't log.